File: tree.hs
import Data.Char
import Data.List

data Tree a =  Node a [Tree a]  deriving (Eq, Show)

t1 = Node 5 [ t2 , t3 , t4]
t2 = Node 6 [Node 8 [Node 7 []]]
t3 = Node 54 [Node 9 []]
t4 = Node 34 []

height (Node a []) = 1
height (Node a ls) = 1 + (maximum . map height $ ls)

sum_tree (Node a []) = a
sum_tree (Node a ls) = a + (sum . map sum_tree $ ls)

-- by using concat and map:

dfs (Node a []) = [a]
dfs (Node a ls) = [a] ++ (concat . map dfs $ ls)

bfs ([]) = []
bfs (ls) =  concat (map children ls) --  (map root ls) ++ bfs (concat (map children ls))

root (Node a ls) = a
children (Node a ls) = ls

-- list comprehension:

dfs2 (Node a []) = [a]
dfs2 (Node a ls) = [a] ++ [ y |  x <- map dfs $ ls , y <- x] 

bfs2 ([]) = []
bfs2 (ls) = (map root ls ) ++ bfs2 [y | x <- map children $ ls , y <- (x)]

File: tree.py
import numpy
import itertools

# dfs of a tree:
def dfs(tree):
 length = len(tree)

 if (not(length == 0)):
   if ((length) == 1):
      return ([tree[0]])
   else:
     return ([tree[0]] + list (numpy.concatenate(map(dfs,tree[1:]))))
 else:
   return []

def dfs2(tree):
 length = len(tree)

 if (not(length == 0)):
   if ((length) == 1):
      return ([tree[0]])
   else:
     return ([tree[0]] + [x for y in map(dfs,tree[1:]) for x in y] )
 else:
   return []


def bfs(tree):
  length = len(tree)
  if (length == 0):
    return []
  else:
    return (map(root,tree) + (bfs(list(itertools.chain.from_iterable(map(children,tree))) )))

def bfs2(tree):
  length = len(tree)
  if (length == 0):
    return []
  else:
    return (map(root,tree) + (bfs([x for y in (map(children,tree)) for x in y]) ))




def root(tree):
   return (tree[0]) 
def children(tree):
   return (tree[1:])

tree = [5,[6,[7],[8,[3],[4],[56],[12],[45],[87]]],[9,[4],[2]]]
print bfs([tree])
print bfs2([tree])


 
